keras/backend/theano_backend.py

```diff
@@ -307,11 +307,15 @@ def moving_average_update(variable, value, momentum):
 
 
 def dot(x, y):
-    # TODO: `keras_shape` inference.
     if is_sparse(x):
-        return th_sparse_module.basic.structured_dot(x, y)
+        out = th_sparse_module.basic.structured_dot(x, y)
     else:
-        return T.dot(x, y)
+        out = T.dot(x, y)
+    if hasattr(x, '_keras_shape') and hasattr(y, '_keras_shape'):
+        np_x = np.zeros(x._keras_shape)
+        np_y = np.zeros(y._keras_shape)
+        out._keras_shape = np.dot(np_x, np_y).shape
```

Comment:
Should we rely have to rely on `np.dot`? Besides the fact that it's expensive, it's also a black box. The logic should be inferable from reading the code.